### ISSUE TITLE
Support `InPlace` upgrade for worker machines

### ISSUE THREAD

**Issue by eromanova** [k0sproject/k0smotron#792]
Currently, according to k0smotron docs, the `InPlace` upgrade (using autopilot) is supported for control plane machines only: https://docs.k0smotron.io/stable/update/update-capi-cluster/. It'll be good to support the same approach for worker machines as well if possible.

**Comment by DinaBelova** [comment]
Dear k0smotron team, can you please share the plan on getting this fixed? This issue was opened some time ago and in general it's important for the k0rdent project experience. (cc @mestadler @p5ntangle fyi folks)

**Comment by jnummelin** [comment]
> can you please share the plan on getting this fixed?

First of all, fix is the wrong word as this is a new feature, not a bug.

I'm not entirely sure if k0smotron can even "fix" this. The worker `MachineDeployment` object is fully controlled by CAPI core controllers and thus very hard to hook into via k0smotron.

From CAPI/k0rdent point of view, how would you want this to work? I.e. what are the expected steps to trigger worker `InPlace` update?

**Comment by eromanova** [comment]
> From CAPI/k0rdent point of view, how would you want this to work? 

From the k0rdent perspective, it would be great to have the same approach we currently use for the `InPlace` upgrade of ControlPlanes. The upgrade strategy (InPlace or Replace) could be configured within one of k0smotron’s public API objects, such as `K0sWorkerConfigTemplate`.

With the `InPlace` upgrade strategy, the node would not be recreated and all updated k0s settings (ex. k0s version) would be applied to the existing node.

I understand that implementing this in k0smotron may be difficult or even impossible, but I wanted to bring it up for discussion (just in case).

**Comment by jnummelin** [comment]
The main challenge is that `K0sWorkerConfigTemplate `, as all other CAPI XYZTemplate types, are not something that are really reconciled by any controller. They are merely "placeholders" when the actual types are created.

We can keep this still around if we come up with any idea how this could be done, but currently it seems pretty impossible to implement.

> I wanted to bring it up for discussion (just in case)
Always good to have discussions around these kinda ideas. 😄

**Comment by kahirokunn** [comment]
This has been merged.
[📖 Add In-place updates proposal by g-gaston · Pull Request #11029 · kubernetes-sigs/cluster-api](https://github.com/kubernetes-sigs/cluster-api/pull/11029)

**Comment by jnummelin** [comment]
Yeah, the _proposal_ has been merged but the feature in CAPI is not yet _implemented_, AFAIK

k0smotron cannot really support this untill  it's really supported in CAPI itself.

**Comment by kahirokunn** [comment]
Oh, okay.

**Comment by pbasov** [comment]
https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/tasks/experimental-features/runtime-sdk/implement-in-place-update-hooks.md
Experimentally out, I suggest we take a look when everyone is back from the holidays

**Comment by apedriza** [comment]
> https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/tasks/experimental-features/runtime-sdk/implement-in-place-update-hooks.md Experimentally out, I suggest we take a look when everyone is back from the holidays

@pbasov we have [this](https://github.com/k0sproject/k0smotron/issues/1307) for more details about support in-place upgrades following CAPI proposal. I will close this issue in favor of that one.
